### PR TITLE
add keycloak id to users table  

### PR DESCRIPTION
**Pull request by davgur**

### CODE REVIEW

File: migrations/2022-02-24_224:44:31_add_kc_to_user.sql [hunk 1]
```diff
@@ -0,0 +1,10 @@
+-- MDB generated migration file
+-- rambler up
+ALTER TABLE users
+    ADD COLUMN accounts_id VARCHAR(36) UNIQUE,
```
Comment (davgur) [inline migrations/2022-02-24_224:44:31_add_kc_to_user.sql:4]: is it must be not null? What will be default unique value?

File: permissions/middleware.go [hunk 2]
```diff
@@ -64,13 +70,57 @@ func AuthenticationMiddleware() gin.HandlerFunc {
 				}
 
 				c.Set("ID_TOKEN_CLAIMS", claims)
+
```
Comment (davgur) [inline permissions/middleware.go:73]: as I understand we want add all users that call mdb, right? Ore make other middleware for labels calls?

File: migrations/2022-02-24_224:44:31_add_kc_to_user.sql (hunk 1 above)
Comment (edoshor) [reply migrations/2022-02-24_224:44:31_add_kc_to_user.sql:4]: I guess you can't add this column in a single step.
You can add it with no constraints and then have another migration to introduce the non-null and uniqueness constraints.

File: permissions/middleware.go [hunk 3]
```diff
@@ -64,13 +70,57 @@ func AuthenticationMiddleware() gin.HandlerFunc {
...
+		Name:       null.StringFrom(fmt.Sprintf("% %", claims.GivenName, claims.FamilyName)),
+	}
+
+	tx, err := mdb.Begin()
+	utils.Must(err)
+	// check if not unique on DB - "23505": "unique_violation",
+	errDB := user.Insert(tx)
+	if errDB != nil {
+		utils.Must(tx.Rollback())
+		if !strings.Contains(errDB.Error(), "pq: duplicate key value violates unique constraint \"users_accounts_id_key\"") {
```
Comment (edoshor) [inline permissions/middleware.go:107]: this is wrong. you should not escape this db error. 
See comment above about 2 step migration

File: permissions/middleware.go [hunk 4]
```diff
@@ -64,13 +70,57 @@ func AuthenticationMiddleware() gin.HandlerFunc {
...
+func getOrCreateUser(mdb *sql.DB, claims *IDTokenClaims) (*models.User, error) {
+	user, err := fetchUserFromDB(mdb, claims.Sub)
+	if err != nil || user != nil {
+		return user, err
+	}
+
+	user = &models.User{
+		AccountsID: null.StringFrom(claims.Sub),
+		Email:      claims.Email,
+		Name:       null.StringFrom(fmt.Sprintf("% %", claims.GivenName, claims.FamilyName)),
```
Comment (edoshor) [inline permissions/middleware.go:98]: this is ok.
also make sure you know strings.Join

File: permissions/middleware.go [hunk 5]
```diff
@@ -64,13 +70,57 @@ func AuthenticationMiddleware() gin.HandlerFunc {
...
 }
 
+func getOrCreateUser(mdb *sql.DB, claims *IDTokenClaims) (*models.User, error) {
+	user, err := fetchUserFromDB(mdb, claims.Sub)
+	if err != nil || user != nil {
+		return user, err
+	}
+
+	user = &models.User{
+		AccountsID: null.StringFrom(claims.Sub),
```
Comment (edoshor) [inline permissions/middleware.go:96]: This should not be null after all migrations are done

File: permissions/middleware.go [hunk 6]
```diff
@@ -64,13 +70,57 @@ func AuthenticationMiddleware() gin.HandlerFunc {
 				}
 
 				c.Set("ID_TOKEN_CLAIMS", claims)
+
+				mdb := c.MustGet("MDB").(*sql.DB)
+				user, err := getOrCreateUser(mdb, &claims)
+				if err != nil {
+					c.AbortWithError(http.StatusBadRequest, err).SetType(gin.ErrorTypePublic)
```
Comment (edoshor) [inline permissions/middleware.go:77]: this is not a 400 bad request error.
it is a 500 internal server error. type private

File: migrations/2022-02-24_224:44:31_add_kc_to_user.sql (hunk 1 above)
Comment (edoshor) [reply migrations/2022-02-24_224:44:31_add_kc_to_user.sql:4]: Between the two steps we should run some script to add `accounts_id` value for all existing users.
You can ask Amnon's help with extracting the data from keycloak.

**Review by edoshor — CHANGES_REQUESTED** [review]

File: migrations/2022-02-24_224:44:31_add_kc_to_user.sql (hunk 1 above)
Comment (davgur) [reply migrations/2022-02-24_224:44:31_add_kc_to_user.sql:4]: you mean that after script that will fill kc id I will add some hardcoded string as default?

Comment (edoshor) [reply migrations/2022-02-24_224:44:31_add_kc_to_user.sql:4]: No. I mean after the script you can add the uniqueness constraint. From there on we enforce the kc ID for every user

File: permissions/middleware.go (hunk 3 above)
Comment (davgur) [reply permissions/middleware.go:107]: i wont check that after line 90 but before line 101 from other thread was not created user

File: importer/keycloak/export_kc.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+			error:    nil,
+		}
+		if err := e.updateUser(u); err != nil {
+			e.resp[i].wasAdded = false
+			e.resp[i].error = err
+			withErr = append(withErr, u)
+			log.Debugf("Error for email: %s Error: %v ", u.Email, err)
+		}
+	}
+	e.useEmailAsAccountId(withErr)
```
Comment (davgur) [inline importer/keycloak/export_kc.go:60]: what you say? have few emails that no on keycloak

Comment (edoshor) [reply importer/keycloak/export_kc.go:60]: I'm fine with that. How many such users we have like this ?

Comment (davgur) [reply importer/keycloak/export_kc.go:60]: 5-6

File: permissions/middleware.go [hunk 8]
```diff
@@ -64,13 +70,53 @@ func AuthenticationMiddleware() gin.HandlerFunc {
...
+
+	tx, err := mdb.Begin()
+	utils.Must(err)
+
+	if err := user.Insert(tx); err != nil {
+		utils.Must(tx.Rollback())
+	} else {
+		utils.Must(tx.Commit())
+	}
+	return fetchUserFromDB(mdb, claims.Sub)
```
Comment (edoshor) [inline permissions/middleware.go:109]: Not sure you have to fetch it from DB here. Consider simply returning `user`.
If you must fetch then there is a simple method on User. `Refresh` or `Reload` can't recall the actual name

File: migrations/2022-02-24_224:44:31_add_kc_to_user.sql [hunk 9]
```diff
@@ -0,0 +1,13 @@
+-- MDB generated migration file
+-- rambler up
+
+ALTER TABLE users
+    ALTER COLUMN name TYPE VARCHAR(255),
+    ADD COLUMN account_id VARCHAR(36),
+    ADD COLUMN disabled   BOOLEAN DEFAULT FALSE NOT NULL;
```
[the remark this answers is not in the document]
Comment (davgur) [reply migrations/2022-02-24_224:44:31_add_kc_to_user.sql:7]: removed

**Review by edoshor — APPROVED** [review]

**Pull request merged**